### PR TITLE
Bridge page: handle bridge errors in UI

### LINKED ISSUE #79: Bridge Page: Decide how to handle various error states on the bridge page
Examples:

* User needs to switch network
* Still needs to set their allowance

Screen Shot 2023-10-25 at 4 26 41 PM

Screen Shot 2023-10-30 at 9 54 27 PM

Possible approach is to disable the simulation until the proper criteria are met (network, allowance, etc..?) and therefore avoid showing or having to suppress error messages that are not useful.

### LINKED ISSUE #81: Bridge page: Handle Insufficient ETH Message Better
* Show near the wormhole relayer field info?
* Make that field look more like a (disabled) input?
* OR maybe show as a modal popup instead?

Screen Shot 2023-10-31 at 11 13 35 AM

### PR DESCRIPTION
**Pull request by wildmolasses**
Resolves #79 
Resolves #81

When not connected:
- don't show any error ✅ 

When connected and on the wrong network:
- don't show any error ✅ 
- Switch network button working. When rejected, show "user rejected request." ✅ 

When connected on correct network:
- don't show any error if amount is 0 ✅ 
- don't show any error if insufficient allowance ✅  (shows set allowance button)
- If not enough voting token balance to bridge, show special erc20 amount error near input ✅ 
- If not enough ETH for relayer fee, show error in error box ✅ 
- can click button when amount is 0; unhandled error ❌  (tracking in #40 and handled in #113)

### CODE REVIEW

**Comment by wildmolasses** [comment]
Current dependencies on/for this PR:
* main
  * **PR #101** <a href="https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/101" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ScopeLift/l2-flexible-voting-frontend/101?utm_source=stack-comment).

File: components/ProposalRow.tsx [hunk 1]
```diff
@@ -128,6 +128,7 @@ export default function ProposalRow({
                       className="rounded bg-indigo-600 px-2 py-1 text-xs font-semibold text-white shadow-sm hover:bg-indigo-500 focus-visible:outline focus-visible:outline-2 focus-visible:outline-offset-2 focus-visible:outline-indigo-600"
```
Comment (garyghayrat) [inline components/ProposalRow.tsx:128]: _(random placement)_

When connected to goerli on the bridge page, clicking `switch direction` causes the error below:
image

Comment (garyghayrat) [reply components/ProposalRow.tsx:128]: Same error if I switch to Optimism Goerli from the network dropdown or if I was already on Optimism Goerli on Metamask

Comment (garyghayrat) [reply components/ProposalRow.tsx:128]: Interesting that this is only happening to one of my tet wallets: 0x637C1Ec1d205a4E7a79c9CE4Bd100CD1d19E6080

File: hooks/useEasyWrite.ts [hunk 2]
```diff
@@ -97,16 +97,12 @@ export const useEasyWrite = (params: UsePrepareContractWriteConfig) => {
     config.request,
```
Comment (garyghayrat) [inline hooks/useEasyWrite.ts:97]: _(random placement)_
Find this a bit jarring, maybe we can keep the error while the amount is being checked and only hide the error once the latest input is correct?

https://github.com/ScopeLift/l2-flexible-voting-frontend/assets/61768337/d39dc47b-ca86-45d5-9b8c-d0667b0d222e

Comment (garyghayrat) [reply hooks/useEasyWrite.ts:97]: Same thing here, also not sure why there's a L shaped white space in the error box.
https://github.com/ScopeLift/l2-flexible-voting-frontend/assets/61768337/57049642-ff46-426d-a23c-2e351dc096c5

File: pages/[id]/bridge.tsx [hunk 3]
```diff
@@ -101,20 +132,66 @@ const Bridge = () => {
...
+      bridgeToL1();
+    }
   };
 
-  const source =
-    bridgeTarget === BridgeTarget.L2
-      ? { ...l1Config, ...l1, fee: fees?.l1 || BigInt(0) }
-      : { ...l2Config, ...l2, fee: fees?.l2 || BigInt(0) };
-  const target =
+  const formatError = (e: Error | null): ErrorReturnType => {
```
Comment (alexkeating) [inline pages/[id]/bridge.tsx:144]: Nit: I think this function could be abstracted out to a utils file.

File: pages/[id]/bridge.tsx [hunk 4]
```diff
@@ -101,20 +132,66 @@ const Bridge = () => {
...
+      },
+      {
+        // Here we duplicate the ERC20AmountError -- this is because the error search string is slightly different for L1 and L2,
+        // but they're functionally equivalent.
+        errorType: ErrorType.ERC20AmountError,
+        errorSearchString: 'ERC20: burn amount exceeds balance',
+        prettyReason: `Error: Not enough ${source.token?.symbol} in wallet.`,
+      },
+      {
+        errorType: ErrorType.ETHError,
```
Comment (alexkeating) [inline pages/[id]/bridge.tsx:165]: This error name is a little vague. Maybe, InsufficientNativeCurrencyError

File: pages/[id]/bridge.tsx [hunk 5]
```diff
@@ -101,20 +132,66 @@ const Bridge = () => {
...
+    if (e === null) return { errorType: undefined, errorReason: undefined };
+    const errorTypes = [
+      {
+        errorType: ErrorType.UserRejectedRequest,
+        errorSearchString: 'User rejected the request.',
+        prettyReason: `Error: User rejected the request.`,
+      },
+      {
+        errorType: ErrorType.ERC20AmountError,
+        errorSearchString: 'ERC20: transfer amount exceeds balance',
```
Comment (alexkeating) [inline pages/[id]/bridge.tsx:154]: Where is the message for the error defined? What if we change the message changes slightly? Will the search then fail for this error.

File: components/ProposalRow.tsx (hunk 1 above)
Comment (wildmolasses) [reply components/ProposalRow.tsx]: thanks @garyghayrat , i ironed out this edge case in latest revision!

File: hooks/useEasyWrite.ts (hunk 2 above)
Comment (wildmolasses) [reply hooks/useEasyWrite.ts:97]: For the jarring error text, i added an `&nbsp;` so the space there exists even when there's no error. should make it less jarring.

can't reproduce the L shaped white space 😵

File: pages/[id]/bridge.tsx (hunk 3 above)
Comment (wildmolasses) [reply pages/[id]/bridge.tsx:159]: Generally agree, but currently this function uses state from the component (aka `source.token?.symbol`, `chain?.nativeCurrency.symbol`, `bridgeToL1Error`, etc. For now i'll leave as is.

File: pages/[id]/bridge.tsx (hunk 4 above)
Comment (wildmolasses) [reply pages/[id]/bridge.tsx]: thanks, done!

File: pages/[id]/bridge.tsx (hunk 5 above)
Comment (wildmolasses) [reply pages/[id]/bridge.tsx:171]: > What if we change the message changes slightly? Will the search then fail for this error.

the message is a viem error. if viem changes their error strings, we'll have to change this error parser. with viem pinned I think we're safe.

**Review by alexkeating — APPROVED** [review]

**Pull request merged**